### PR TITLE
feat(balances): transfers & rewards

### PR DESCRIPTION
**Pull request by deployAt**
- [x] Transfer bets to jackblock pot
- [x] Calculate balance rewards
- [x] Transfer rewards to winners + fees

### CODE REVIEW

File: _POC/winner/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+      (w2, 3),
+      (w3, 2),
+      (w4, 5),
+      (w5, 6),
+      (w6, 4),
+    ];
+
+    // filter out "w3" because hits < 3
+    let winners_to_reward: Winners<AccountId> = winners.into_iter().filter(|&(_, hits) | hits >= 3).collect();
+
```
Comment (deployAt) [inline _POC/winner/src/main.rs:43]: no way to to these computations in one chain 😕
https://stackoverflow.com/questions/33189227/how-to-get-a-slice-from-an-iterator

File: pallets/jackblock/src/lib.rs [hunk 2]
```diff
@@ -124,13 +135,19 @@ decl_storage! {
...
 
 decl_event!(
-	pub enum Event<T> where AccountId = <T as frame_system::Config>::AccountId {
+	pub enum Event<T> where
+		AccountId = <T as frame_system::Config>::AccountId,
+		Balance = BalanceOf<T>,
+		{
 		NewBet(SessionIdType, Bet<AccountId>),
 		Winners(SessionIdType, Winners<AccountId>),
+		RewardFeeForAuthority(AccountId, Balance),
```
Comment (korzewski) [inline pallets/jackblock/src/lib.rs:149]: To be consistent can we rename it to RewardFeeForAuthorities?

File: pallets/jackblock/src/lib.rs [hunk 3]
```diff
@@ -199,15 +221,85 @@ decl_module! {
...
+				let authorities = Self::authorities();
+				let authorities_count = authorities.len() as u32;
+				let reward_fee_per_authority: BalanceOf<T> = fees / authorities_count.into(); // TODO: fixed point safe division
+
+				debug::info!("--- Pot before: {:?}", pot);
+				debug::info!("--- Pot for fees: {:?} $", fees);
+				debug::info!("--- Pot for rewards: {:?} $", pot_for_rewards);
+
+				for authoritiy in authorities {
+					debug::info!("--- Reward for authoricity: {:?}, {:?} $", authoritiy, reward_fee_per_authority);
```
Comment (korzewski) [inline pallets/jackblock/src/lib.rs:242]: Just a typo, "Reward for authority"

File: pallets/jackblock/src/lib.rs [hunk 4]
```diff
@@ -199,15 +221,85 @@ decl_module! {
...
+
+	fn distribute_reward(reward_percentage: u8, winners: &[(Bet<T::AccountId>, u8)], pot_for_rewards: BalanceOf<T>, hits: u8) {
+		let rewards_from_pot = Percent::from_percent(reward_percentage) * pot_for_rewards;
+		let winners_count = winners.len() as u32;
+		let reward_per_winner: BalanceOf<T> = rewards_from_pot / winners_count.into(); // TODO: fixed point safe division
+
+		winners.iter().for_each(|winner| {
+			let winner_account = &winner.0.account_id;
+			debug::info!("Account {:?} won {:?} $ by guessing {:?} numbers!", winner_account, reward_per_winner, hits);
+			let _ = T::Currency::transfer(&Self::account_id(), &winner_account, reward_per_winner, KeepAlive);
```
Comment (korzewski) [inline pallets/jackblock/src/lib.rs:298]: Error handling by `?` is missing at the end of this line

File: pallets/jackblock/src/lib.rs [hunk 5]
```diff
@@ -283,11 +375,11 @@ impl<T: Config> Module<T> {
 			let next_session_number = Self::get_random_number();
 			if !session_numbers.contains(&next_session_number) {
 				session_numbers[i] = next_session_number;
-			    i += 1;
+					i += 1;
```
Comment (korzewski) [inline pallets/jackblock/src/lib.rs:378]: I think this extra indent is not necessary here

File: pallets/jackblock/src/lib.rs [hunk 6]
```diff
@@ -283,11 +375,11 @@ impl<T: Config> Module<T> {
...
 			}
-			
+
 			if i == GUESS_NUMBERS_COUNT {
-			    break;
+					break;
```
Comment (korzewski) [inline pallets/jackblock/src/lib.rs:382]: I think this extra indent is not necessary here

**Review by korzewski — COMMENTED** [review]
Everything looks great!

I've just added a few minor typos etc

File: pallets/jackblock/src/lib.rs (hunk 4 above)
Comment (deployAt) [reply pallets/jackblock/src/lib.rs]: will create a task for it

File: pallets/jackblock/src/lib.rs (hunk 2 above)
Comment (deployAt) [reply pallets/jackblock/src/lib.rs:149]: but it emits an event for a single authority

**Review by korzewski — APPROVED** [review]

**Pull request merged**